src/QSymbolicsBase/predefined.jl

```diff
@@ -207,25 +229,94 @@ function Base.show(io::IO, x::SProjector)
     print(io,"]")
 end
 
-"""Dagger a Ket into Bra."""
-@withmetadata struct SDagger <: Symbolic{AbstractBra}
-    ket::Symbolic{AbstractKet}
+"""Dagger, i.e., adjoint of quantum objects (kets, bras, operators)
+
+```jldoctest 
+julia> a = SKet(:a, SpinBasis(1//2)); A = SOperator(:A, SpinBasis(1//2));
+
+julia> dagger(2*im*A*a)
+0 - 2im⟨a|A†
+
+julia> B = SOperator(:B, SpinBasis(1//2));
+
+julia> dagger(A*B)
+B†A†
+
+julia> ℋ = SHermitianOperator(:ℋ, SpinBasis(1//2)); U = SUnitaryOperator(:U, SpinBasis(1//2));
+
+julia> dagger(ℋ)
+ℋ
+
+julia> dagger(U) 
+U⁻¹
+```
+"""
+@withmetadata struct SDagger{T<:QObj} <: Symbolic{T}
+    obj
 end
 isexpr(::SDagger) = true
 iscall(::SDagger) = true
-arguments(x::SDagger) = [x.ket]
+arguments(x::SDagger) = [x.obj]
 operation(x::SDagger) = dagger
 head(x::SDagger) = :dagger
-children(x::SDagger) = [:dagger, x.ket]
-dagger(x::Symbolic{AbstractKet}) = SDagger(x)
-dagger(x::SScaledKet) = SScaledBra(x.coeff, dagger(x.obj))
+children(x::SDagger) = [:dagger, x.obj]
+dagger(x::Symbolic{AbstractBra}) = SDagger{AbstractKet}(x)
+dagger(x::Symbolic{AbstractKet}) = SDagger{AbstractBra}(x)
+dagger(x::Symbolic{AbstractOperator}) = SDagger{AbstractOperator}(x)
+dagger(x::SKet) = SBra(x.name, x.basis)
```

Comment:
OK, suppose we have a symbolic ket, say `|a⟩`, and we want to apply `dagger` to this ket. So instead of following the definition that `dagger(|a⟩) == ⟨a|`, as done in this PR, we should leave it as `dagger(|a⟩) == |a⟩†`? And this is because we are kind of cheating with the former approach, rather than traversing the expression tree? I'm not disagreeing with you, I'm just making sure I'm understanding what you're saying.

Reply:
Yup, that is it. There is no guarantee that `|a⟩` and `⟨a|` are related at all (except for you implicitly deciding so). To make it explicit you need to write some pattern matching rule that makes it evident that they are related. Such a pattern matching rule would be difficult in a setup that is not based on the expressions API. By just creating a new expression type `dagger`, that problem of implicit relationships disappears -- everything becomes explicit.

Reply:
Got it, just made changes.